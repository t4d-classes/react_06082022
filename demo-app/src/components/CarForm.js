import { useState } from 'react';

export const CarForm = ({ buttonText, onSubmitCar }) => {

  const [ carForm, setCarForm ] = useState({
    make: '', model: '', year: 1900, color: '', price: 0,
  });

  const change = (evt) => {
    setCarForm({
      ...carForm,
      [evt.target.name]: evt.target.type === 'number'
        ? evt.target.valueAsNumber : evt.target.value,
    });
  };

  const submitCar = () => {

    onSubmitCar({ ...carForm });

    setCarForm({
      make: '', model: '', year: 1900, color: '', price: 0,
    });

  };

  return (
    <form>
      <label>
        Make:
        <input type="text" name="make" value={carForm.make} onChange={change} />
      </label>
      <label>
        Model:
        <input type="text" name="model" value={carForm.model} onChange={change} />
      </label>
      <label>
        Year:
        <input type="number" name="year" value={carForm.year} onChange={change} />
      </label>
      <label>
        Color:
        <input type="text" name="color" value={carForm.color} onChange={change} />
      </label>
      <label>
        Price:
        <input type="number" name="price" value={carForm.price} onChange={change} />
      </label>
      <button type="button" onClick={submitCar}>{buttonText}</button>
    </form>
  );

};

CarForm.defaultProps = {
  buttonText: 'Submit Car',
};