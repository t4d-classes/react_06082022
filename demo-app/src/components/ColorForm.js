import { useState } from 'react';

export const ColorForm = ({ buttonText }) => {

  const [ colorForm, setColorForm ] = useState({
    name: '', hexcode: '',
  });

  const change = (evt) => {
    setColorForm({
      ...colorForm,
      [evt.target.name]: evt.target.value,
    });
  };

  console.log(colorForm);

  return (
    <form>
      <label>
        Name:
        <input type="text" name="name" value={colorForm.name} onChange={change} />
      </label>
      <label>
        Hexcode:
        <input type="text" name="hexcode" value={colorForm.hexcode} onChange={change} />
      </label>
      <button type="button">{buttonText}</button>
    </form>
  );

};

ColorForm.defaultProps = {
  buttonText: 'Submit Color',
};