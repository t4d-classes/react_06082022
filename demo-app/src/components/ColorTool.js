import PropTypes from 'prop-types';

export const ColorTool = (props) => {

  return (
    <>
      <header>
        <h1>Color Tool</h1>
      </header>
      <ul>
        {props.colors.map(color => <li key={color.id}>{color.name}</li>)}
      </ul>
    </>
  );

};

ColorTool.defaultProps = {
  colors: [],
};

ColorTool.propTypes = {
  colors: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    hexcode: PropTypes.string.isRequired,
  })).isRequired,
};