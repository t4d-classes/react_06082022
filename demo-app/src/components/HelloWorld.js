// custom functional component
export const HelloWorld = () => {

  // this is not HTML, this is JSX (is not valid JavaScript)
  // JSX is a function call
  // return React.createElement('h1', null, 'Hello, World!')React.createElement('span', null, 'test');
  return (
    <>
      <h1>Hello, World!</h1>
      <span>test</span>
    </>
  );

};