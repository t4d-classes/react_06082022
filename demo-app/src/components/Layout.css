/* http://meyerweb.com/eric/tools/css/reset/ 
   v2.0 | 20110126
   License: none (public domain)
*/

html, body, div, span, applet, object, iframe,
h1, h2, h3, h4, h5, h6, p, blockquote, pre,
a, abbr, acronym, address, big, cite, code,
del, dfn, em, img, ins, kbd, q, s, samp,
small, strike, strong, sub, sup, tt, var,
b, u, i, center,
dl, dt, dd, ol, ul, li,
fieldset, form, label, legend,
table, caption, tbody, tfoot, thead, tr, th, td,
article, aside, canvas, details, embed, 
figure, figcaption, footer, header, hgroup, 
menu, nav, output, ruby, section, summary,
time, mark, audio, video {
	margin: 0;
	padding: 0;
	border: 0;
	font-size: 100%;
	font: inherit;
	vertical-align: baseline;
}
/* HTML5 display-role reset for older browsers */
article, aside, details, figcaption, figure, 
footer, header, hgroup, menu, nav, section {
	display: block;
}
body {
	line-height: 1;
}
ol, ul {
	list-style: none;
}
blockquote, q {
	quotes: none;
}
blockquote:before, blockquote:after,
q:before, q:after {
	content: '';
	content: none;
}
table {
	border-collapse: collapse;
	border-spacing: 0;
}

body {
    font-size: 100%;
    font-family: Arial, Helvetica, sans-serif;
}

.container {

    position: fixed;
    top: 0; bottom: 0;
    left: 0; right: 0;

    display: grid;

    grid-template-rows: 4rem 2rem 1fr 2rem;
    grid-template-columns: 4fr 1fr;

    grid-template-areas: 
        "header header"
        "menu menu"
        "content sidebar"
        "footer footer";
}

#page-header, #page-footer {
    background-color: navy;
    color: white;
}

#page-header > h1 {
  font-size: 2rem;
}

#page-header {
    grid-area: header;
    display: flex;
    flex-direction: row;
    justify-content: flex-start; /* major axis */
    align-items: center; /* minor axis */
    padding-left: 20px;
}

#main-menu {
    grid-area: menu;
    background-color: lightgray;
    padding: 0 20px;
    display: flex;
    flex-direction: row;
    align-items: center;
}

/* select the ul that is a child of the #main-menu */
#main-menu > ul {
    display: flex;
    flex-direction: row;
    list-style-type: none;
    padding: 0;
    margin: 0;
}

/* select the li that is a descendant of the #main-menu */
#main-menu li {
    margin: 4px;
}

#main-menu li > a,
#main-menu li > a:visited,
#main-menu li > a:active {
    color: black;
    text-decoration: none;
}

#main-menu li > a:hover {
    text-decoration: underline;
}


#content {
    grid-area: content;
    padding: 10px;
}

#sidebar {
    grid-area: sidebar;
    background-color: orange;
    padding: 10px;
}

#page-footer {
    grid-area: footer;
    font-size: 0.7rem;
    text-align: center;
    padding-top: 6px;
}