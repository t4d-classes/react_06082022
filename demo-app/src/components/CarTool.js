import { useState } from 'react';

import { ToolHeader } from './ToolHeader';
import { CarTable } from './CarTable';
import { CarForm } from './CarForm';

export const CarTool = ({ cars: initialCars }) => {

  const [ cars, setCars ] = useState([...initialCars]);

  const addCar = (car) => {
    setCars([
      ...cars,
      {
        ...car,
        id: Math.max(...cars.map(c => c.id), 0) + 1,
      },
    ]);
  };

  return (
    <>
      <ToolHeader headerText="Car Tool" />
      <CarTable cars={cars} />
      <CarForm buttonText="Add Car" onSubmitCar={addCar} />
    </>
  );

};

CarTool.defaultProps = {
  cars: [],
};