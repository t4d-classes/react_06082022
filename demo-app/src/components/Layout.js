import { Link, Outlet } from 'react-router-dom';

import "./Layout.css";

export const Layout = () => {

  return (
    <div className="container">

      <header id="page-header">
        <h1>Tools App</h1>
      </header>

      <nav id="main-menu">
        <ul>
          <li><Link to="/">Home</Link></li>
          <li><Link to="/color-tool">Color Tool</Link></li>
          <li><Link to="/car-tool">Car Tool</Link></li>
        </ul>
      </nav>

      <main id="content">
        <Outlet />
      </main>

      <aside id="sidebar">
        sidebar
      </aside>

      <footer id="page-footer">
        <small>&copy; 2022 A Cool Company, Inc.</small>
      </footer>

    </div>
  )




};