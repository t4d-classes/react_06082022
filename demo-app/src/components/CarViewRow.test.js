import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

import { CarViewRow } from '../components/CarViewRow';

test('snapshot CarViewRow component', () => {
  const car = {
    id: 1,
    make: 'Ford',
    model: 'Fusion Hybrid',
    year: 2020,
    color: 'red',
    price: 45000,
  };

  expect(
    renderer
      .create(
        <CarViewRow
          car={car}
          onEditCar={() => null}
          onDeleteCar={() => null}
        />,
      )
      .toJSON(),
  ).toMatchSnapshot();
});

describe('CarViewRow testing library', () => {
  let car;
  let carKeys;
  let editCarSpy;
  let deleteCarSpy;

  beforeEach(() => {
    car = {
      id: 1,
      make: 'Ford',
      model: 'Fusion Hybrid',
      year: 2020,
      color: 'red',
      formattedPrice: '45000',
      price: 45000,
    };

    carKeys = Object.keys(car);

    editCarSpy = jest.fn();
    deleteCarSpy = jest.fn();
  });

  const renderComponent = () => {
    render(
      <table>
        <tbody>
          <CarViewRow
            car={car}
            onEditCar={editCarSpy}
            onDeleteCar={deleteCarSpy}
          />
        </tbody>
      </table>,
    )    
  };

  test('render CarViewRow component', () => {

    renderComponent();
    
    // <td> elements have a role of "cell"
    const cells = screen.getAllByRole('cell');
    expect(cells.length).toBe(7);
    cells.slice(0, 6).forEach((element, index) => {
      expect(element.textContent).toBe(String(car[carKeys[index]]));
    });
  });

  test('click edit', () => {

    renderComponent();
    
    // select the button element by text
    userEvent.click(screen.getByText('Edit'));
    expect(editCarSpy).toHaveBeenCalledWith(1);
  });

  test('click delete', () => {

    renderComponent();
    
    // select the button element by text
    userEvent.click(screen.getByText('Delete'));
    expect(deleteCarSpy).toHaveBeenCalledWith(1);
  });
});
