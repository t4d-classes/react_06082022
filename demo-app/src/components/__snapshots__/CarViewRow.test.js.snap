// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot CarViewRow component 1`] = `
<tr>
  <td>
    1
  </td>
  <td>
    Ford
  </td>
  <td>
    Fusion Hybrid
  </td>
  <td>
    2020
  </td>
  <td>
    red
  </td>
  <td>
    45000
  </td>
  <td>
    <button
      onClick={[Function]}
    >
      Edit
    </button>
    <button
      onClick={[Function]}
    >
      Delete
    </button>
  </td>
</tr>
`;
